tests/unit/test_transaction.py

```diff
@@ -672,6 +823,332 @@ def test_context_mgr_failure(self):
+        from .test_snapshot import _MockIterator
+
+        # mock transaction and data process
+        VALUES = [[u"bharney", u"rhubbyl", 31], [u"phred", u"phlyntstone", 32]]
+        VALUE_PBS = [[_make_value_pb(item) for item in row] for row in VALUES]
+
+        SQL_QUERY_WITH_PARAM = """
+SELECT first_name, last_name, email FROM citizens WHERE age <= @max_age"""
+        PARAMS = {"max_age": 30}
+        PARAM_TYPES = {"max_age": "INT64"}
+
+        struct_type_pb = StructType(
+            fields=[
+                StructType.Field(name="first_name", type=Type(code=STRING)),
+                StructType.Field(name="last_name", type=Type(code=STRING)),
+                StructType.Field(name="age", type=Type(code=INT64)),
+            ]
+        )
+        database = _Database()
+        transaction = self._make_one(_Session(database))
+
+        result_sets = [
+            PartialResultSet(
+                values=VALUE_PBS[0], metadata=ResultSetMetadata(row_type=struct_type_pb)
+            ),
+            PartialResultSet(values=VALUE_PBS[1]),
+        ]
+        now_pb = _datetime_to_pb_timestamp(
+            datetime.datetime.utcnow().replace(tzinfo=UTC)
+        )
+        api = database.spanner_api = _FauxSpannerAPI(
+            _begin_transaction_response=TransactionPB(id=self.TRANSACTION_ID),
+            _commit_response=CommitResponse(commit_timestamp=now_pb),
+            _rollback_response=Empty(),
+        )
+        api.execute_streaming_sql = mock.Mock(return_value=_MockIterator(*result_sets))
+
+        transaction._original_results_checksum = ResultsChecksum()
+        transaction._original_results_checksum.consume_result(7)
+
+        # run SQL query
+        with self.assertRaises(RuntimeError):
+            with transaction:
+                list(
+                    transaction.execute_sql(
+                        SQL_QUERY_WITH_PARAM, PARAMS, PARAM_TYPES, query_mode=2
+                    )
+                )
+
+
+class TestResultsChecksum(unittest.TestCase):
+    def _getTargetClass(self):
+        from google.cloud.spanner_v1.transaction import ResultsChecksum
+
+        return ResultsChecksum
+
+    def _make_one(self):
+        return self._getTargetClass()()
+
+    def test_consume_result_row(self):
+        ETALON_CHECKSUM = b"\x822\xea\x0ep\xfd\xe6%\x01!\xe6\xb0A?\x11k\xdc\xde\x0c\x0cn\xbcH\x1e\xdf\x12\xb7\xd1:\x1d\x08\xfc"
+
+        checksum = self._make_one()
+        checksum.consume_result(["val1", "val2"])
+
+        self.assertEqual(checksum.checksum.digest(), ETALON_CHECKSUM)
+        self.assertEqual(checksum.count, 1)
+
+    def test_consume_result_int(self):
+        ETALON_CHECKSUM = b"l\x0e\xb4\xa6\xdeeU\xa3.\x85\x93u-\xee \xd1B\x1a\x96\xfd\xed\x96\x1a\xb3y\xb5\xb7\x12a\x0e4\xec"
+
+        checksum = self._make_one()
+        checksum.consume_result(5)
+
+        self.assertEqual(checksum.checksum.digest(), ETALON_CHECKSUM)
+        self.assertEqual(checksum.count, 1)
```

Comment:
I think using `pickle` for the serialization is going to make these tests pretty brittle. It looks like they pass on 3.6 and 3.7 but fail on 3.8:

```
% python --version
Python 3.8.2
% python -m py.test tests/unit/test_transaction.py
============================================================================================ test session starts ============================================================================================
platform darwin -- Python 3.8.2, pytest-6.0.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/libc/src/python-spanner
collected 50 items

tests/unit/test_transaction.py ................................................FF                                                                                                                     [100%]

================================================================================================= FAILURES ==================================================================================================
________________________________________________________________________________ TestResultsChecksum.test_consume_result_int ________________________________________________________________________________

self = <tests.unit.test_transaction.TestResultsChecksum testMethod=test_consume_result_int>

    def test_consume_result_int(self):
        ETALON_CHECKSUM = b"l\x0e\xb4\xa6\xdeeU\xa3.\x85\x93u-\xee \xd1B\x1a\x96\xfd\xed\x96\x1a\xb3y\xb5\xb7\x12a\x0e4\xec"

        checksum = self._make_one()
        checksum.consume_result(5)

>       self.assertEqual(checksum.checksum.digest(), ETALON_CHECKSUM)
E       AssertionError: b'\x84\xbat/\x02\x7f\tn"\xb3\x17\xd9\xd4\x8[42 chars]x0ex' != b'l\x0e\xb4\xa6\xdeeU\xa3.\x85\x93u-\xee \x[50 chars]\xec'

tests/unit/test_transaction.py:1125: AssertionError
________________________________________________________________________________ TestResultsChecksum.test_consume_result_row ________________________________________________________________________________

self = <tests.unit.test_transaction.TestResultsChecksum testMethod=test_consume_result_row>

    def test_consume_result_row(self):
        ETALON_CHECKSUM = b"\x822\xea\x0ep\xfd\xe6%\x01!\xe6\xb0A?\x11k\xdc\xde\x0c\x0cn\xbcH\x1e\xdf\x12\xb7\xd1:\x1d\x08\xfc"

        checksum = self._make_one()
        checksum.consume_result(["val1", "val2"])

>       self.assertEqual(checksum.checksum.digest(), ETALON_CHECKSUM)
E       AssertionError: b'\x10\xe7F\xb4\x9cfb\xc6\xe5\x0f\x189R\xf1\x[43 chars]xca0' != b'\x822\xea\x0ep\xfd\xe6%\x01!\xe6\xb0A?\x11k[51 chars]\xfc'

tests/unit/test_transaction.py:1116: AssertionError
========================================================================================== short test summary info ==========================================================================================
FAILED tests/unit/test_transaction.py::TestResultsChecksum::test_consume_result_int - AssertionError: b'\x84\xbat/\x02\x7f\tn"\xb3\x17\xd9\xd4\x8[42 chars]x0ex' != b'l\x0e\xb4\xa6\xdeeU\xa3.\x85\x93u-\x...
FAILED tests/unit/test_transaction.py::TestResultsChecksum::test_consume_result_row - AssertionError: b'\x10\xe7F\xb4\x9cfb\xc6\xe5\x0f\x189R\xf1\x[43 chars]xca0' != b'\x822\xea\x0ep\xfd\xe6%\x01!\xe6\x...
======================================================================================= 2 failed, 48 passed in 1.10s ========================================================================================
```

One (admittedly ugly) option is to call `pickle.dumps` in the tests here too instead of the checksum constants.

Reply:
Yeah, that's something I was internally ready for, just didn't thought it'll fail on such a small amounts of data. Reworked